### DISCUSSION TITLE
Handle reauthentication prompts/redirects from authenticating reverse proxies like Entra App Proxy gracefully (without a background CORS error only to be fixed by shift+F5)

### DISCUSSION THREAD

**Discussion by JasperE84** [open-webui/open-webui#7714] [Issues]
I have made open-webui available for my organisation using Entra Application Proxy. This is more or less Microsofts version of a cloudflared tunnel: WAN facing web requests are terminated at the microsoft cloud, which prompts the user with an Entra authentication request. This setup allows me to safely expose the app to WAN, by inserting an Entra auth gateway in front of it, allowing full conditional access (incl. device based access restrictions), and it allows me to realize SSO, seamless sign on, and role based access control to the web app.

Header based auth is realized by configuring the env var `WEBUI_AUTH_TRUSTED_EMAIL_HEADER` and then providing that header through the Entra App Proxy.  

In effect opening https://openwebui.mydomain.com will result in an entra authentication dialog (which is handled automatically if seamless sign on is configured in intune), and will then reverse-proxy to the internally hosted web app. The user will then be instantly logged in in open-webui.

This all works perfectly except for one little aspect: If the web app makes an API call, and the entra proxy requires a reauth, the api call to openwebui.mydomain.com, gets redirected to the login.microsoftonline.com domain. This will then result in a CORS error in the browser console. This initially happened each time I restarted my browser and a new sessions is made, making entra ask for reauth. What I think happens, is the open-webui app is cached in the browser, and it immediately makes the API call, resulting in an error because openwebui.mydomain.com is not in the microsoft CORS list, and an authentication page is returned instead of API data. A shift-f5 in the browser will then promptly fix that, but that requires user knowledge. To improve this I have set the entra reverse proxy cookie to persistent, stopping Entra from redirecting to login.microsoftonline.com, making it work almost perfectly. Yet I still occasionally get a 500 error (gray background in the open-webui app with a lightgray 500 error). Before I made the entra cookie persistent, I also sometimes got an error screen in with a message saying something like 'Oops. You're using a non supported method, host the webui from the backend' (translated to english), with a "Check again" button. (This check-again button however wouldnt initiate a shift-f5 like action, so the reload didn't work)

This would be fixed if Open-WebUI can handle/detect such redirections (to login.microsoftonline.com in this case) and force-reload the entire app if this happens. Or make a button on the 500 error page which will do this.

I have tried playing with the JWT token expiry but that caused similar errors (having to manually refresh the browser once expired)

Here's my `.env` config:
```
WEBUI_URL=https://openwebui.mydomain.com
OPEN_WEBUI_PORT=3031
WEBUI_AUTH=true
WEBUI_AUTH_TRUSTED_EMAIL_HEADER=X-User-OpenWebUi-Email
ENABLE_SIGNUP=true
DEFAULT_USER_ROLE=user
JWT_EXPIRES_IN=-1
```

And Entra settings:
![image](https://github.com/user-attachments/assets/9553ed38-a972-4492-bc4f-1f8b3ab01cc6)
![image](https://github.com/user-attachments/assets/1d10ca84-4454-47c2-8780-7fd358a1bc83)
![image](https://github.com/user-attachments/assets/e418caf7-e49d-4103-9a14-51a22e11f404)

**Comment by tjbck** [comment]
Help from the community wanted here, keep us updated!

**Comment by JasperE84** [comment]
Some more details: Here's whats happening.

The console log referring to layout.svelte:106
![image](https://github.com/user-attachments/assets/c5b82037-eaf2-4af9-87ca-7d4d3e4ca3f9)

layout.svelte:106
![image](https://github.com/user-attachments/assets/4316a73c-f969-4f03-bba1-409d27978076)

Meanwhile in the UI:
![image](https://github.com/user-attachments/assets/d96889c6-1442-43fd-97d8-178e49f3594d)

UI served from cache:
![image](https://github.com/user-attachments/assets/6d5d2548-b47d-4cec-86ca-185dda64922e)

But config request API call gets redirected
![image](https://github.com/user-attachments/assets/59c12362-5e1f-414c-af3a-b56d92478795)
![image](https://github.com/user-attachments/assets/8acb6489-73c8-4812-93d2-4adf112b92f2)

Fails to fetch due to CORS error
![image](https://github.com/user-attachments/assets/308ce52c-bdc1-48d8-8cf6-47ef67882756)
![image](https://github.com/user-attachments/assets/671f9c1b-62ca-419a-b73a-82635f2051cf)

**Comment by JasperE84** [comment]
Just another thought about a solution; can cache be disabled on the https://openwebui.mydomain.com/ root url? If that root URL wouldn't be cached, it would be redirected over login.microsoftonline.com if the session was invalidated by Entra, and the auth would work. 

Unfortunately can't seem to figure out how to mod the code to achieve this.

**EDIT: Strike that, I figured out some (rather ugly) method to force cache invalidation on the root url.**

The Entra reverse proxy is now pointing to a dockerized nginx reverse proxy, which is pointing to the actual open-webui webapp.

Here's the nginx reverse proxy config to invalidate cache on the root URL.
First testing seems ok now, can disable the "use persistent cookies" in entra now. Fingers crossed.

```
server {
  listen 3032;

  location / {
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "Upgrade";
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;

    # Invalidate cache for root url to support login.microsoftonline.com redirections
    add_header Cache-Control "no-store, no-cache, must-revalidate, proxy-revalidate, max-age=0";
    add_header Pragma "no-cache";
    add_header Expires "0";


    proxy_read_timeout 86400;

    proxy_pass http://svrdocker01.stl.local:3031;
  }

````

**Reply by gomble** [reply]
> Just another thought about a solution; can cache be disabled on the https://openwebui.mydomain.com/ root url? If that root URL wouldn't be cached, it would be redirected over login.microsoftonline.com if the session was invalidated by Entra, and the auth would work.
> 
> Unfortunately can't seem to figure out how to mod the code to achieve this.
> 
> **EDIT: Strike that, I figured out some (rather ugly) method to force cache invalidation on the root url.**
> 
> The Entra reverse proxy is now pointing to a dockerized nginx reverse proxy, which is pointing to the actual open-webui webapp.
> 
> Here's the nginx reverse proxy config to invalidate cache on the root URL. First testing seems ok now, can disable the "use persistent cookies" in entra now. Fingers crossed.
> 
> ```
> server {
>   listen 3032;
> 
>   location / {
>     proxy_http_version 1.1;
>     proxy_set_header Upgrade $http_upgrade;
>     proxy_set_header Connection "Upgrade";
>     proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
>     proxy_set_header X-Forwarded-Proto $scheme;
> 
>     # Invalidate cache for root url to support login.microsoftonline.com redirections
>     add_header Cache-Control "no-store, no-cache, must-revalidate, proxy-revalidate, max-age=0";
>     add_header Pragma "no-cache";
>     add_header Expires "0";
> 
> 
>     proxy_read_timeout 86400;
> 
>     proxy_pass http://svrdocker01.stl.local:3031;
>   }
> ```

i have tested this but getting the 500 error:
Do you found a solution to this ?

**Comment by rohanhzri** [comment]
Any update @JasperE84 ?

**Comment by idle-loop** [comment]
Hey, was there any update or more info on this? Thanks.

**Comment by IOITJ** [comment]
Having the same issues with Entra App Proxy. Would be good to have a solution for this.

**Comment by BobbyD4AI** [comment]
Just found this whilst researching how to secure OWUI in an Azure environment... Entra App Proxy looks like a plausible route but before I go down that route it'd be good to do if any solutions have been found? Thanks.

**Comment by JasperE84** [comment]
I have since switched to another LLM chat interface where I faced similar problems. I posted a workaround which might also work for open-webui here: https://github.com/danny-avila/LibreChat/discussions/8051
